File: Test/17_api_validation/test_api_validation.R
# CASSIA Test 17: API Key Validation Testing (R)
# ================================================
# Tests the validate_api_keys() function for different LLM providers.
#
# This test verifies:
# - API key validation for OpenAI, Anthropic, and OpenRouter
# - Caching behavior (second validation should be instant)
# - Force revalidation
# - Error handling for invalid/missing keys
#
# Usage:
#     Rscript test_api_validation.R

# Get script directory (works with Rscript and source())
get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg), winslash = "/")))
  }
  for (i in sys.nframe():1) {
    if (!is.null(sys.frame(i)$ofile)) {
      return(dirname(normalizePath(sys.frame(i)$ofile, winslash = "/")))
    }
  }
  return(normalizePath(getwd(), winslash = "/"))
}
script_dir <- get_script_dir()

# Source shared utilities
source(file.path(script_dir, "..", "shared", "r", "test_utils.R"))
source(file.path(script_dir, "..", "shared", "r", "result_manager.R"))
source(file.path(script_dir, "..", "shared", "r", "logging_manager.R"))

run_api_validation_test <- function() {
  print_test_header("17 - API Key Validation Testing (R)")

  # Load configuration
  config <- load_config()
  print_config_summary(config)

  # Setup API keys
  setup_api_keys()

  # Setup CASSIA R package (for Python interop via reticulate)
  tryCatch({
    setup_cassia()
    message("CASSIA R package loaded successfully")
  }, error = function(e) {
    message("Error loading CASSIA: ", e$message)
    return(FALSE)
  })

  # Import CASSIA
  tryCatch({
    CASSIA <- import_cassia()
    message("CASSIA Python module imported successfully")
  }, error = function(e) {
    message("Error importing CASSIA: ", e$message)
    return(FALSE)
  })

  # Create results directory
  results_dir <- create_results_dir("17_api_validation")
  message(sprintf("Results will be saved to: %s", results_dir))

  start_time <- Sys.time()
  errors <- list()
  all_passed <- TRUE

  # Test 1: Validate all providers
  cat("\n", strrep("=", 60), "\n")
  cat("Test 1: Validate All Providers\n")
  cat(strrep("=", 60), "\n")

  tryCatch({
    result <- CASSIA$validate_api_keys(verbose=TRUE)
    message("Validation result: ", toString(result))

    if (is.null(result) || (!is.logical(result) && length(result) == 0)) {
      message("  [SKIP] No API keys configured")
    } else if (is.logical(result)) {
      if (result) {
        message("  [OK] API key is valid")
      } else {
        message("  [FAIL] API key is invalid")
        all_passed <- FALSE
      }
    } else {
      # Dictionary result
      message("  [OK] Validated multiple providers")
    }
  }, error = function(e) {
    message("  [ERROR] ", e$message)
    errors <- append(errors, sprintf("validate_all: %s", e$message))
    all_passed <- FALSE
  })

  # Test 2: Validate specific provider
  cat("\n", strrep("=", 60), "\n")
  cat("Test 2: Validate Specific Provider (OpenAI)\n")
  cat(strrep("=", 60), "\n")

  tryCatch({
    if (Sys.getenv("OPENAI_API_KEY") != "") {
      result <- CASSIA$validate_api_keys(provider="openai", verbose=TRUE)
      message("OpenAI validation: ", result)

      if (result) {
        message("  [OK] OpenAI key is valid")
      } else {
        message("  [FAIL] OpenAI key is invalid")
        all_passed <- FALSE
      }
    } else {
      message("  [SKIP] OPENAI_API_KEY not set")
    }
  }, error = function(e) {
    message("  [ERROR] ", e$message)
    errors <- append(errors, sprintf("validate_openai: %s", e$message))
    all_passed <- FALSE
  })

  # Test 3: Test caching behavior
  cat("\n", strrep("=", 60), "\n")
  cat("Test 3: Caching Behavior\n")
  cat(strrep("=", 60), "\n")

  tryCatch({
    if (Sys.getenv("OPENAI_API_KEY") != "") {
      # Clear cache
      CASSIA$clear_validation_cache(provider="openai")

      # First validation
      message("First validation (should make API call)...")
      start1 <- Sys.time()
      result1 <- CASSIA$validate_api_keys(provider="openai", verbose=FALSE)
      duration1 <- as.numeric(difftime(Sys.time(), start1, units="secs"))
      message(sprintf("  Duration: %.3f seconds", duration1))

      # Second validation (should use cache)
      message("Second validation (should use cache)...")
      start2 <- Sys.time()
      result2 <- CASSIA$validate_api_keys(provider="openai", verbose=FALSE)
      duration2 <- as.numeric(difftime(Sys.time(), start2, units="secs"))
      message(sprintf("  Duration: %.6f seconds", duration2))

      if (duration2 < 0.01) {
        message("  [OK] Cache is working (second call < 10ms)")
      } else {
        message("  [WARN] Cache may not be working optimally")
      }
    } else {
      message("  [SKIP] OPENAI_API_KEY not set")
    }
  }, error = function(e) {
    message("  [ERROR] ", e$message)
    errors <- append(errors, sprintf("caching: %s", e$message))
  })

  # Test 4: Invalid API key
  cat("\n", strrep("=", 60), "\n")
  cat("Test 4: Invalid API Key\n")
  cat(strrep("=", 60), "\n")

  tryCatch({
    fake_key <- "sk-fake-key-for-testing-12345"
    result <- CASSIA$validate_api_keys(provider="openai", api_key=fake_key, verbose=TRUE)

    if (!result) {
      message("  [OK] Invalid key correctly detected")
    } else {
      message("  [FAIL] Invalid key reported as valid!")
      all_passed <- FALSE
    }
  }, error = function(e) {
    # Exception is also acceptable for invalid key
    message("  [OK] Exception raised for invalid key")
  })

  duration <- as.numeric(difftime(Sys.time(), start_time, units="secs"))

  # Summary
  cat("\n", strrep("=", 60), "\n")
  cat("SUMMARY\n")
  cat(strrep("=", 60), "\n")
  cat(sprintf("  Status: %s\n", if (all_passed) "PASSED" else "FAILED"))
  cat(sprintf("  Duration: %.2f seconds\n", duration))
  cat(sprintf("  Errors: %d\n", length(errors)))

  # Save metadata
  metadata <- create_test_metadata(
    test_name = "api_validation_test_r",
    config = config,
    duration_seconds = duration,
    status = if (all_passed) "passed" else "failed",
    clusters_tested = list(),
    errors = errors
  )
  save_test_metadata(results_dir, metadata)

  # Print final result
  print_test_result(all_passed, sprintf("Duration: %.2f seconds", duration))

  return(all_passed)
}

# Run the test
success <- tryCatch({
  run_api_validation_test()
}, error = function(e) {
  message("Test failed with error: ", e$message)
  FALSE
})

# Exit with appropriate code
quit(status = if (success) 0 else 1)

File: Test/18_api_validation/test_api_validation.py
"""
CASSIA Test 18: API Key Validation Testing
==========================================
Tests the validate_api_keys() function for different LLM providers.

This test verifies:
- API key validation for OpenAI, Anthropic, and OpenRouter
- Caching behavior (second validation should be instant)
- Force revalidation
- Cache invalidation on key change
- Thread safety for concurrent validations
- Error handling for invalid/missing keys

Usage:
    python test_api_validation.py
"""

import os
import sys
import time
import threading
from pathlib import Path

# Add shared utilities to path
sys.path.insert(0, str(Path(__file__).parent.parent / "shared" / "python"))

from test_utils import (
    setup_cassia_imports,
    load_config,
    setup_api_keys,
    print_test_header,
    print_test_result,
    print_config_summary
)
from result_manager import (
    create_results_dir,
    save_test_metadata,
    save_test_results,
    create_test_metadata
)

# Setup CASSIA imports
setup_cassia_imports()

# Try to load local API keys file if it exists
try:
    import set_api_keys  # Load API keys from local file
    print("Loaded API keys from set_api_keys.py")
except ImportError:
    pass  # No local keys file, will use setup_api_keys() instead


def test_validation_basic():
    """Test basic API key validation for all providers."""
    from CASSIA import validate_api_keys, clear_validation_cache

    print("\n" + "=" * 60)
    print("Test 1: Basic API Key Validation")
    print("=" * 60)

    results = {
        "test_name": "basic_validation",
        "providers": []
    }

    # Test each provider that has a key set
    providers = ["openai", "anthropic", "openrouter"]
    env_vars = {
        "openai": "OPENAI_API_KEY",
        "anthropic": "ANTHROPIC_API_KEY",
        "openrouter": "OPENROUTER_API_KEY"
    }

    for provider in providers:
        print(f"\nTesting {provider}...")
        env_var = env_vars[provider]

        if not os.environ.get(env_var):
            print(f"  [SKIP] {env_var} not set")
            results["providers"].append({
                "provider": provider,
                "status": "skipped",
                "reason": "API key not set"
            })
            continue

        try:
            start = time.time()
            is_valid = validate_api_keys(provider, verbose=True)
            duration = time.time() - start

            results["providers"].append({
                "provider": provider,
                "status": "passed" if is_valid else "failed",
                "valid": is_valid,
                "duration_seconds": duration
            })

            print(f"  Result: {'VALID' if is_valid else 'INVALID'}")
            print(f"  Duration: {duration:.3f}s")

        except Exception as e:
            results["providers"].append({
                "provider": provider,
                "status": "error",
                "error": str(e)
            })
            print(f"  [ERROR] {str(e)[:100]}")

    return results


def test_caching_behavior():
    """Test that validation caching works correctly."""
    from CASSIA import validate_api_keys, clear_validation_cache

    print("\n" + "=" * 60)
    print("Test 2: Caching Behavior")
    print("=" * 60)

    # Find a provider with API key set
    providers = ["openai", "anthropic", "openrouter"]
    env_vars = {
        "openai": "OPENAI_API_KEY",
        "anthropic": "ANTHROPIC_API_KEY",
        "openrouter": "OPENROUTER_API_KEY"
    }

    test_provider = None
    for provider in providers:
        if os.environ.get(env_vars[provider]):
            test_provider = provider
            break

    if not test_provider:
        print("  [SKIP] No API keys configured")
        return {"test_name": "caching_behavior", "status": "skipped"}

    print(f"\nUsing provider: {test_provider}")

    # Clear cache first
    clear_validation_cache(test_provider)

    # First validation (should make API call)
    print("\nFirst validation (should make API call)...")
    start1 = time.time()
    result1 = validate_api_keys(test_provider, verbose=True)
    duration1 = time.time() - start1
    print(f"  Duration: {duration1:.3f}s")

    # Second validation (should use cache)
    print("\nSecond validation (should use cache)...")
    start2 = time.time()
    result2 = validate_api_keys(test_provider, verbose=True)
    duration2 = time.time() - start2
    print(f"  Duration: {duration2:.3f}s")

    # Check that second call was much faster (< 10ms)
    speedup = duration1 / duration2 if duration2 > 0 else float('inf')
    cache_worked = duration2 < 0.01  # Less than 10ms

    print(f"\nCache performance:")
    print(f"  First call:  {duration1:.3f}s (API call)")
    print(f"  Second call: {duration2:.6f}s (cached)")
    print(f"  Speedup: {speedup:.0f}x")
    print(f"  Cache working: {'YES' if cache_worked else 'NO'}")

    return {
        "test_name": "caching_behavior",
        "provider": test_provider,
        "first_duration": duration1,
        "second_duration": duration2,
        "speedup": speedup,
        "cache_working": cache_worked,
        "status": "passed" if cache_worked else "failed"
    }


def test_force_revalidate():
    """Test force_revalidate parameter."""
    from CASSIA import validate_api_keys, clear_validation_cache

    print("\n" + "=" * 60)
    print("Test 3: Force Revalidation")
    print("=" * 60)

    # Find a provider with API key set
    providers = ["openai", "anthropic", "openrouter"]
    env_vars = {
        "openai": "OPENAI_API_KEY",
        "anthropic": "ANTHROPIC_API_KEY",
        "openrouter": "OPENROUTER_API_KEY"
    }

    test_provider = None
    for provider in providers:
        if os.environ.get(env_vars[provider]):
            test_provider = provider
            break

    if not test_provider:
        print("  [SKIP] No API keys configured")
        return {"test_name": "force_revalidate", "status": "skipped"}

    print(f"\nUsing provider: {test_provider}")

    # First validation
    print("\nFirst validation...")
    validate_api_keys(test_provider, verbose=False)

    # Second validation with force_revalidate=True (should make API call)
    print("\nForced revalidation (should make API call even though cached)...")
    start = time.time()
    result = validate_api_keys(test_provider, force_revalidate=True, verbose=True)
    duration = time.time() - start

    # Should take longer than cache lookup (> 100ms typically)
    api_call_made = duration > 0.1

    print(f"  Duration: {duration:.3f}s")
    print(f"  API call made: {'YES' if api_call_made else 'NO (used cache)'}")

    return {
        "test_name": "force_revalidate",
        "provider": test_provider,
        "duration": duration,
        "api_call_made": api_call_made,
        "status": "passed" if api_call_made else "failed"
    }


def test_invalid_key():
    """Test validation with invalid API key."""
    from CASSIA import validate_api_keys

    print("\n" + "=" * 60)
    print("Test 4: Invalid API Key")
    print("=" * 60)

    print("\nTesting with fake API key...")

    # Use a fake key that will fail
    fake_key = "sk-fake-key-for-testing-12345"

    try:
        result = validate_api_keys("openai", api_key=fake_key, verbose=True)

        if not result:
            print("  [OK] Invalid key correctly detected")
            status = "passed"
        else:
            print("  [FAIL] Invalid key reported as valid!")
            status = "failed"

    except Exception as e:
        # Exception is also acceptable
        print(f"  [OK] Exception raised: {str(e)[:80]}")
        status = "passed"

    return {
        "test_name": "invalid_key",
        "status": status
    }


def test_missing_key():
    """Test validation when API key is not set."""
    from CASSIA import validate_api_keys

    print("\n" + "=" * 60)
    print("Test 5: Missing API Key")
    print("=" * 60)

    # Save original key
    original_key = os.environ.get("OPENAI_API_KEY")

    try:
        # Temporarily remove the key
        if "OPENAI_API_KEY" in os.environ:
            del os.environ["OPENAI_API_KEY"]

        print("\nTesting with no API key set...")
        result = validate_api_keys("openai", verbose=True)

        if not result:
            print("  [OK] Missing key correctly detected")
            status = "passed"
        else:
            print("  [FAIL] Missing key reported as valid!")
            status = "failed"

    finally:
        # Restore original key
        if original_key:
            os.environ["OPENAI_API_KEY"] = original_key

    return {
        "test_name": "missing_key",
        "status": status
    }


def test_thread_safety():
    """Test concurrent validation calls (thread safety)."""
    from CASSIA import validate_api_keys, clear_validation_cache

    print("\n" + "=" * 60)
    print("Test 6: Thread Safety")
    print("=" * 60)

    # Find a provider with API key set
    providers = ["openai", "anthropic", "openrouter"]
    env_vars = {
        "openai": "OPENAI_API_KEY",
        "anthropic": "ANTHROPIC_API_KEY",
        "openrouter": "OPENROUTER_API_KEY"
    }

    test_provider = None
    for provider in providers:
        if os.environ.get(env_vars[provider]):
            test_provider = provider
            break

    if not test_provider:
        print("  [SKIP] No API keys configured")
        return {"test_name": "thread_safety", "status": "skipped"}

    print(f"\nUsing provider: {test_provider}")

    # Clear cache first
    clear_validation_cache(test_provider)

    print("\nRunning 5 concurrent validations...")

    results_list = []
    threads = []

    def validate_thread():
        result = validate_api_keys(test_provider, verbose=False)
        results_list.append(result)

    start = time.time()

    # Create and start threads
    for i in range(5):
        thread = threading.Thread(target=validate_thread)
        threads.append(thread)
        thread.start()

    # Wait for all threads
    for thread in threads:
        thread.join()

    duration = time.time() - start

    # Check that all returned the same result
    all_same = len(set(results_list)) == 1
    all_valid = all(results_list)

    print(f"  Results: {results_list}")
    print(f"  All consistent: {'YES' if all_same else 'NO'}")
    print(f"  All valid: {'YES' if all_valid else 'NO'}")
    print(f"  Duration: {duration:.3f}s")

    status = "passed" if (all_same and all_valid) else "failed"

    return {
        "test_name": "thread_safety",
        "provider": test_provider,
        "results": results_list,
        "all_consistent": all_same,
        "all_valid": all_valid,
        "duration": duration,
        "status": status
    }


def run_api_validation_test():
    """Run all API validation tests."""
    print_test_header("18 - API Key Validation Testing")

    # Load configuration
    config = load_config()
    print_config_summary(config)

    # Setup API keys
    setup_api_keys()

    # Create results directory
    results_dir_dict = create_results_dir("18_api_validation")
    results_dir = results_dir_dict['base']  # Extract the base path
    print(f"\nResults will be saved to: {results_dir}")

    start_time = time.time()
    all_test_results = {}
    errors = []

    # Run all tests
    try:
        all_test_results["basic_validation"] = test_validation_basic()
    except Exception as e:
        errors.append(f"basic_validation: {str(e)}")
        print(f"\n[ERROR] Basic validation test failed: {str(e)}")

    try:
        all_test_results["caching_behavior"] = test_caching_behavior()
    except Exception as e:
        errors.append(f"caching_behavior: {str(e)}")
        print(f"\n[ERROR] Caching test failed: {str(e)}")

    try:
        all_test_results["force_revalidate"] = test_force_revalidate()
    except Exception as e:
        errors.append(f"force_revalidate: {str(e)}")
        print(f"\n[ERROR] Force revalidate test failed: {str(e)}")

    try:
        all_test_results["invalid_key"] = test_invalid_key()
    except Exception as e:
        errors.append(f"invalid_key: {str(e)}")
        print(f"\n[ERROR] Invalid key test failed: {str(e)}")

    try:
        all_test_results["missing_key"] = test_missing_key()
    except Exception as e:
        errors.append(f"missing_key: {str(e)}")
        print(f"\n[ERROR] Missing key test failed: {str(e)}")

    try:
        all_test_results["thread_safety"] = test_thread_safety()
    except Exception as e:
        errors.append(f"thread_safety: {str(e)}")
        print(f"\n[ERROR] Thread safety test failed: {str(e)}")

    duration = time.time() - start_time

    # Calculate summary
    total_tests = len(all_test_results)
    passed_tests = sum(1 for r in all_test_results.values()
                      if isinstance(r, dict) and r.get("status") == "passed")
    failed_tests = sum(1 for r in all_test_results.values()
                      if isinstance(r, dict) and r.get("status") == "failed")
    skipped_tests = sum(1 for r in all_test_results.values()
                       if isinstance(r, dict) and r.get("status") == "skipped")

    # Print summary
    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)
    print(f"  Passed:  {passed_tests}/{total_tests}")
    print(f"  Failed:  {failed_tests}/{total_tests}")
    print(f"  Skipped: {skipped_tests}/{total_tests}")
    print(f"  Duration: {duration:.2f}s")

    # Determine overall status
    if passed_tests == total_tests:
        status = "passed"
    elif passed_tests > 0 and failed_tests == 0:
        status = "passed_with_skipped"
    else:
        status = "failed"

    # Save metadata and results
    metadata = create_test_metadata(
        test_name="api_validation_test",
        config=config,
        duration_seconds=duration,
        status=status,
        clusters_tested=[],
        errors=errors
    )
    metadata["test_summary"] = {
        "passed": passed_tests,
        "failed": failed_tests,
        "skipped": skipped_tests,
        "total": total_tests
    }
    save_test_metadata(results_dir, metadata)
    save_test_results(results_dir, all_test_results)

    # Print final result
    success = (status == "passed" or status == "passed_with_skipped")
    print_test_result(success, f"Duration: {duration:.2f}s")

    return success


if __name__ == "__main__":
    success = run_api_validation_test()
    sys.exit(0 if success else 1)
